pkg/agent/externalnode/external_node_controller.go

```diff
@@ -0,0 +1,757 @@
+		return fmt.Errorf("failed to reconcile reserved flows %v", err)
+	}
+	klog.Info("ExternalNodeController finishes reconciliation")
+	return nil
+}
+
+func (c *ExternalNodeController) reconcileHostUplinkFlows() error {
+	hostIfaces := c.ifaceStore.GetInterfacesByType(interfacestore.ExternalEntityInterface)
+	for _, hostIface := range hostIfaces {
+		if err := c.ofClient.InstallVMUplinkFlows(hostIface.InterfaceName, hostIface.OVSPortConfig.OFPort, hostIface.UplinkPort.OFPort); err != nil {
+			klog.ErrorS(err, "Failed to install host uplink flows", "hostInterface", hostIface.InterfaceName)
+			return err
+		}
+		klog.InfoS("Reconciled host uplink flow for ExternalEntityInterface", "ifName", hostIface.InterfaceName)
+	}
+	return nil
+}
+
+func (c *ExternalNodeController) reconcileExternalNodeInterfaces() error {
+	en, err := c.externalNodeLister.ExternalNodes(c.namespace).Get(c.nodeName)
+	if err != nil {
+		return err
+	}
+	if err = c.addExternalNode(en); err != nil {
+		return err
+	}
+	eeName, err := externalnode.GenExternalEntityName(en)
+	if err != nil {
+		return err
+	}
+	hostIfaces := c.ifaceStore.GetInterfacesByType(interfacestore.ExternalEntityInterface)
+	for _, hostIface := range hostIfaces {
+		if hostIface.EntityName != eeName || hostIface.EntityNamespace != en.Namespace {
+			if err = c.deleteInterface(hostIface.InterfaceName); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *ExternalNodeController) reconcilePolicyBypassFlows() error {
+	rules, err := c.getPolicyBypassRules()
+	if err != nil {
+		return err
+	}
+	for _, rhp := range rules {
+		klog.V(2).InfoS("Installing policy bypass flows", "protocol", rhp.protocol, "IP", rhp.ip, "IPNet", rhp.ipnet, "port", rhp.port, "ingress", rhp.ingress)
+		if err = c.ofClient.InstallPolicyBypassFlows(rhp.protocol, rhp.ipnet, rhp.ip, rhp.port, rhp.ingress); err != nil {
+			klog.ErrorS(err, "Failed to install policy bypass flows", "protocol", rhp.protocol, "IP", rhp.ip, "IPNet", rhp.ipnet, "port", rhp.port, "direction", rhp.ingress)
+			return err
+		}
+	}
+	klog.InfoS("Installed policy bypass flows", "RuleCount", len(rules))
+	return nil
+}
+
+func (c *ExternalNodeController) getPolicyBypassRules() ([]policyBypassRule, error) {
+	rules := make([]policyBypassRule, 0)
+	for _, s := range c.policyBypassRules {
+		// Each rule contains the following four attributes:
+		// direction (in|out), protocol(tcp/udp/icmp), remote IP or CIDR, dst port (ICMP doesn't require),
+		// and these attributes are separated by ":", e.g. "in:tcp:1.1.1.1:22".
+		rule := strings.Split(s, policyBypassRulesSplitter)
+
+		// parse direction.
+		direction := false
+		if rule[0] == "in" {
+			direction = true
+		}
+
+		// parse protocol.
+		var proto binding.Protocol
+		switch rule[1] {
+		case "tcp":
+			proto = binding.ProtocolTCP
+		case "udp":
+			proto = binding.ProtocolUDP
+		case "icmp":
+			proto = binding.ProtocolICMP
+		default:
+			proto = binding.ProtocolIP
```

Comment:
I feel we should either define a name for IP or default to IP if no protocol is specified, but should return an error when validating the rules if not supported protocol is added.